From 987adb0334bce877f0eb7bbf0dfef07d2f9917ec Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <thomas.langer@intel.com>
Date: Fri, 10 Jan 2020 19:45:27 +0100
Subject: [PATCH] PONRTSYS-6161: Add support for relocation from high to low
 addresses

Some of the checks assumed that the relocation is always executed
from low to high addresses. Correct this to allow a negative offset
for the relocation.
---
 arch/mips/kernel/relocate.c | 23 ++++++++++++++++-------
 arch/mips/kernel/setup.c    |  7 +++++--
 2 files changed, 21 insertions(+), 9 deletions(-)

diff --git a/arch/mips/kernel/relocate.c b/arch/mips/kernel/relocate.c
index 6d1271837ae3..c54f41e960d6 100644
--- a/arch/mips/kernel/relocate.c
+++ b/arch/mips/kernel/relocate.c
@@ -289,15 +289,24 @@ static inline void __init *determine_relocation_address(void)
 
 static inline int __init relocation_addr_valid(void *loc_new)
 {
+	unsigned long kernel_length;
+
 	if ((unsigned long)loc_new & 0x0000ffff) {
 		/* Inappropriately aligned new location */
 		return 0;
 	}
-	if ((unsigned long)loc_new < (unsigned long)&_end) {
-		/* New location overlaps original kernel */
-		return 0;
-	}
-	return 1;
+	kernel_length = (long)_end - (long)(&_text);
+
+	/* enough space before kernel */
+	if ((unsigned long)loc_new + kernel_length < (unsigned long)&_text)
+		return 1;
+
+	/* behind the original kernel */
+	if ((unsigned long)loc_new > (unsigned long)&_end)
+		return 1;
+
+	/* New location overlaps original kernel */
+	return 0;
 }
 
 void *__init relocate_kernel(void)
@@ -371,11 +380,11 @@ void *__init relocate_kernel(void)
  */
 void show_kernel_relocation(const char *level)
 {
-	unsigned long offset;
+	long offset;
 
 	offset = __pa_symbol(_text) - __pa_symbol(VMLINUX_LOAD_ADDRESS);
 
-	if (IS_ENABLED(CONFIG_RELOCATABLE) && offset > 0) {
+	if (IS_ENABLED(CONFIG_RELOCATABLE) && offset != 0) {
 		printk(level);
 		pr_cont("Kernel relocated by 0x%pK\n", (void *)offset);
 		pr_cont(" .text @ 0x%pK\n", _text);
diff --git a/arch/mips/kernel/setup.c b/arch/mips/kernel/setup.c
index da1bc450439e..b911c31fc89c 100644
--- a/arch/mips/kernel/setup.c
+++ b/arch/mips/kernel/setup.c
@@ -598,19 +598,22 @@ static void __init bootmem_init(void)
 	 */
 	if (__pa_symbol(_text) > __pa_symbol(VMLINUX_LOAD_ADDRESS)) {
 		unsigned long offset;
-		extern void show_kernel_relocation(const char *level);
 
 		offset = __pa_symbol(_text) - __pa_symbol(VMLINUX_LOAD_ADDRESS);
 		free_bootmem(__pa_symbol(VMLINUX_LOAD_ADDRESS), offset);
+	}
 
 #if defined(CONFIG_DEBUG_KERNEL) && defined(CONFIG_DEBUG_INFO)
+	{
 		/*
 		 * This information is necessary when debugging the kernel
 		 * But is a security vulnerability otherwise!
 		 */
+		extern void show_kernel_relocation(const char *level);
 		show_kernel_relocation(KERN_INFO);
-#endif
 	}
+#endif
+
 #endif
 
 	/*
